common/src/warp_runner/manager/commands/constellation_commands.rs

```diff
@@ -464,31 +474,37 @@ async fn handle_upload_progress(
     warp_storage: &mut warp_storage,
     mut upload_progress: ConstellationProgressStream,
     filename: String,
+    file_id: Uuid,
+    file_state: TransferState,
     file_path: PathBuf,
 ) {
     let tx_upload_file = UPLOAD_FILE_LISTENER.tx.clone();
     let mut previous_percentage: usize = 0;
     let mut upload_process_started = false;
+    let mut last_progress = None;
 
     while let Some(upload_progress) = upload_progress.next().await {
+        last_progress = Some(upload_progress.clone());
+        let current_progress = upload_progress.clone();
         match upload_progress {
             Progression::CurrentProgress {
                 name,
                 current,
                 total,
             } => {
                 log::trace!("starting upload file action listener");
-                if let Ok(received_tx) = CANCEL_FILE_UPLOADLISTENER
-                    .rx
-                    .clone()
-                    .lock()
-                    .await
-                    .try_recv()
-                {
-                    if received_tx {
-                        let _ = tx_upload_file.send(UploadFileAction::Cancelling);
-                        break;
+                let mut first = true;
+                while file_state.matches(TransferStates::Pause).await {
```

Comment:
What we could do instead is use `select` macro from either futures or tokio and send a signal to pause the polling of the stream.

Reply:
hmm not too sure what you mean by that.
you mean replacing it with something like this?
```rust
tokio::select! {
  v = file_state.matches(TransferStates::Pause) => if v { sleep(Duration::from_secs(1)).await },
}
```

Reply:
Something along those lines, although we wouldnt need to sleep or anything. Unless we wish to impl our own stream surrounding the state of the transfer (which might more desirable to do but not really needed), we can add the stream to the select branch and be biased about the order of which things are polled. Eg it would be something like the following 

```rust
    loop {
        tokio::select! {
            biased;
            _ = file_state.matches(TransferStates::Cancel) => return,
            _ = file_state.matches(TransferStates::Pause) => {
                tokio::time::sleep(Duration::from_secs(1)).await;
            },
            upload_progress = upload_progress.next() => {
                let Some(progress) = upload_progress else {
                    break;
                };
                match progress {
                    Progression::CurrentProgress { name, current, total } => {
                        if !upload_process_started {
                            upload_process_started = true;
                            log::info!("Starting upload for {name}");
                            log::info!("0% completed -> written {current} bytes")
                        };

                        if let Some(total_size) = total {
                            let current_percentage = (((current as f64) / (total_size as f64)) * 100.) as usize;
                            if previous_percentage != current_percentage {
                                previous_percentage = current_percentage;
                                let readable_current = format_size(current, DECIMAL);
                                let percentage_number = ((current as f64) / (total_size as f64)) * 100.;
                                let _ = tx_upload_file.send(UploadFileAction::Uploading((
                                    Some(Progression::CurrentProgress { name, current, total }), //format!("{}%", percentage_number as usize),
                                    get_local_text("files.uploading-file"),
                                    file_id,
                                )));
                                log::info!(
                                    "{}% completed -> written {readable_current}",
                                    percentage_number as usize
                                )
                            }
                            // ConstellationProgressStream only ends (atm) when all files in the queue are done uploading
                            // This causes pending file count to not be updated which is way we send a message here too
                            if current_percentage == 100 {
                                let _ = tx_upload_file.send(UploadFileAction::Finishing(
                                    file_path.clone(),
                                    file_id,
                                    false,
                                ));
                            }
                        }
                    },
                    Progression::ProgressComplete { name, total } => {
                        let total_size = total.unwrap_or_default();
                        let readable_total = format_size(total_size, DECIMAL);
                        let _ = tx_upload_file.send(UploadFileAction::Uploading((
                            Some(Progression::ProgressComplete { name: name.clone(), total }), //"100%".into(),
                            get_local_text("files.uploading-file"),
                            file_id,
                        )));
                        log::info!("{name} has been uploaded with {}", readable_total);
                    },
                    Progression::ProgressFailed { name, last_size, error } => {
                        log::info!(
                            "{name} failed to upload at {} MB due to: {}",
                            last_size.unwrap_or_default(),
                            error.unwrap_or_default()
                        );
                        let _ = tx_upload_file.send(UploadFileAction::Error(
                            Some(file_path.clone()),
                            Some(file_id),
                        ));
                        break;
                    },
                }
            }
        }
    }

```

Do note this is just from me eyeing it but it would be something like this.

Reply:
hmm im not too convinced about that.
we need to do something like this:
```rust
    let mut paused = false;

    loop {
        tokio::select! {
            biased;
            _ = async {}, if file_state.matches(TransferStates::Cancel).await => break,
            _ = async {}, if file_state.matches(TransferStates::Pause).await => {
                if !paused {
                    let _ = tx_upload_file.send(UploadFileAction::Pausing(file_id));
                    paused = true;
                }
            },
            upload_progress = upload_progress.next() => {
                paused = false;
                let Some(upload_progress) = upload_progress else {
                    break;
                };
                last_progress = Some(upload_progress.clone());
                let current_progress = upload_progress.clone();
                match upload_progress {
                    Progression::CurrentProgress {
[...]
```

as `file_state.matches` returns an async bool that should decide whether that section is run. (your version e.g. would always run the first task)

so this whole thing just feels kinda bloated

Reply:
actually no. as my version
`_ = async {}, if file_state.matches(TransferStates::Cancel).await => break,`
 will run that task if the state is cancel
but your version
`v = file_state.matches(TransferStates::Cancel) => break`
will run it regardless of the state as there is no condition to stop that.
and having something like 
`v = file_state.matches(TransferStates::Cancel), if v => break` will not work as its not supported

Reply:
Keep in mind i stated the code I provided was an example (since I was eyeing it from github). I am aware that my example doesnt check the boolean there (which was intentional). I was more referring to the stream itself. On top of that, i did mention in my previous statement that we could have the state local so it could be checked or changed there when attempting to poll the stream by doing `upload_progress = async {}, matches!(state, TransferStates::Normal) => {}` (or if you wish to continue to use `files_state` you can use that in place of the match macro) or something along those lines.

Reply:
if we want to have the state local though we would need some sort of channel that tells it to set the value.
and when i initially designed this i opted against channels cause:
1. the channel needs to be either global and when sending e.g. an id that matches the current stream over that channel we update the value but this would mean every stream would need to listen to that channel and i also am not sure what happens if they all try to poll some value there
2. if the channel is also local we would need to send the channel itself back to our tracker which in that case would be just better to keep the current impl

3. maybe something like this could work?
```rust
    let mut paused = false;

    loop {
        let state = file_state.get().await;
        tokio::select! {
            biased;
            _ = async {}, if matches!(state, TransferStates::Cancel).await => break,
            _ = async {}, if matches!(state, TransferStates::Pause).await => {
                if !paused {
                    let _ = tx_upload_file.send(UploadFileAction::Pausing(file_id));
                    paused = true;
                }
            },
            upload_progress = upload_progress.next() => {
                paused = false;
                let Some(upload_progress) = upload_progress else {
                    break;
                };
                last_progress = Some(upload_progress.clone());
                let current_progress = upload_progress.clone();
                match upload_progress {
                    Progression::CurrentProgress {
[...]
```
but like i said i feel like this whole thing has not much of a point